File: 2017.4.1/getEdgePosition.m
function [x_edge, y_edge] = getEdgePosition(img, block_size, thr)
%get all edge laplace
%============================================
%Author:Chame
%Date:2017.3.28
%============================================
%Input
% img: input image
% block_size: 1/2 size of block to determine edge;
% thr: if value > thr, it will be edge;
%--------------------------------------------
%Output
% x_edge:the positions of horizontal edges
% y_edge:the positions of vertical edges
%============================================
narginchk(1, 3);
if nargin == 1
    block_size = 19;
    thr = (2*block_size+1)^2 / 2 * 0.2;
elseif nargin == 2
    thr = (2*block_size+1)^2 / 2 * 0.2;
end
[m, n] = size(img);
h_x = ones(block_size * 2 + 1);
h_y = ones(block_size * 2 + 1);
h_x(:, 1:block_size) = -1;
h_y(block_size+2:end,:) = -1;
value_x = imfilter(img, h_x, 'symmetric');
value_y = imfilter(img, h_y, 'symmetric');

result_x = value_x > thr;
result_y = value_y > thr;
[indy, indx] = find(result_x);
for i = 1:size(indy, 1)
    if indy(i)-block_size < 1 || indy(i)+block_size > m || indx(i)-block_size < 1 || indx(i)+block_size > n
        result_x(indy(i), indx(i)) = 0;
    else
        temp = value_x(indy(i)-block_size:indy(i)+block_size, indx(i)-block_size:indx(i)+block_size);
        [max_y, max_x] = find(temp == max(temp(:)));
        max_y = max_y(1) + indy(i)-block_size - 1;
        max_x = max_x(1) + indx(i)-block_size - 1;
        if indy(i) ~=  max_y || indx(i) ~= max_x
            result_x(indy(i), indx(i)) = 0;           
        end
    end
end
[indy, indx] = find(result_y);
for i = 1:size(indy, 1)
    if indy(i)-block_size < 1 || indy(i)+block_size > m || indx(i)-block_size < 1 || indx(i)+block_size > n
        result_y(indy(i), indx(i)) = 0;
    else
        temp = value_y(indy(i)-block_size:indy(i)+block_size, indx(i)-block_size:indx(i)+block_size);
        [max_y, max_x] = find(temp == max(temp(:)));
        max_y = max_y(1) + indy(i)-block_size - 1;
        max_x = max_x(1) + indx(i)-block_size - 1;
        if indy(i) ~= max_y(1) || indx(i) ~= max_x(1)
            result_y(indy(i), indx(i)) = 0;           
        end
    end
end

[indy,indx] = find(result_x);
x_edge(:, 1) = indx;
x_edge(:, 2) = indy;
[indy,indx] = find(result_y);
y_edge(:, 1) = indx;
y_edge(:, 2) = indy;

figure;
imshow(img, []);
hold on;
plot(x_edge(:, 1), x_edge(:, 2), '.', 'color', 'y');
for i = 1:size(x_edge, 1)
    rectangle('Position', [x_edge(i, 1) - block_size, x_edge(i, 2) - block_size, 2*block_size+1, 2*block_size+1], 'EdgeColor','r');
end
plot(y_edge(:, 1), y_edge(:, 2), '.', 'color', 'r');
for i = 1:size(y_edge, 1)
    rectangle('Position', [y_edge(i, 1) - block_size, y_edge(i, 2) - block_size, 2*block_size+1, 2*block_size+1], 'EdgeColor','r');
end

end

File: V2_C++Code/getEdgePosition.cpp
#include <opencv2/core/core.hpp> 
#include <opencv2/highgui/highgui.hpp> 
#include <opencv2/imgproc/imgproc.hpp> 
#include <opencv2/features2d/features2d.hpp>
#include <iostream>
#include <math.h>

using namespace std;
using namespace cv;

typedef struct {
	vector<Point> x_edge;
	vector<Point> y_edge;
}HV_Edge; // get horizontal and vertical edge positions

HV_Edge getEdgePosition(Mat img, int block_size, float thr);
/*
%get all edge position
%============================================
%Author:Chame & Ocean(Get the pixel value using pointer)
%Date : 2017.4.10
%============================================
%Input
% img : input image
% block_size : 1 / 2 size of block to determine edge;
% thr: if value > thr, it will be edge;
%--------------------------------------------
%Output
% x_edge:the positions of horizontal edges
% y_edge : the positions of vertical edges
%============================================
*/
HV_Edge getEdgePosition(Mat img, int block_size, float thr)
{
	int height = img.rows;
	int width = img.cols;
	//定义模板
	Mat h_x = Mat::ones(block_size * 2 + 1, block_size * 2 + 1, CV_32F);
	h_x.rowRange(0, block_size - 1) = -1;
	Mat h_y = Mat::ones(block_size * 2 + 1, block_size * 2 + 1, CV_32F);
	h_y.colRange(0, block_size - 1) = -1;
	Mat value_x, value_y;
	Point Pblock_size; Pblock_size.x = block_size; Pblock_size.y = block_size;
	//通过模板做相关运算
	filter2D(img, value_x, img.depth(), h_x, Point(-1, -1), (0, 0), BORDER_REFLECT);
	filter2D(img, value_y, img.depth(), h_y, Point(-1, -1), (0, 0), BORDER_REFLECT);
	//找到相关性大于阈值的的点的位置，并做非极大抑制，确定刃边的位置
	HV_Edge edge;
	for (int j = block_size ; j < height - block_size ; j++)
	{
		float* data_x = value_x.ptr<float>(j);
		float* data_y = value_y.ptr<float>(j);
		for (int i = block_size ; i < width - block_size ; i++)
		{
			if (data_x[i] > thr)
			{
				Mat roi(value_x, Rect(i - block_size, j - block_size, 2 * block_size + 1, 2 * block_size + 1));
				Point maxPoint;
				minMaxLoc(roi, 0, 0, 0, &maxPoint);
				maxPoint.x = maxPoint.x + i - block_size;
				maxPoint.y = maxPoint.y + j - block_size;
				if ( i == maxPoint.x && j == maxPoint.y ){
					edge.x_edge.push_back(maxPoint);
					//rectangle(img, maxPoint - Pblock_size, maxPoint + Pblock_size, cvScalar(255, 0, 0), 2);
				}
				
			}
			if (data_y[i] > thr)
			{
				Mat roi(value_y, Rect(i - block_size, j - block_size, 2 * block_size + 1, 2 * block_size + 1));
				Point maxPoint;
				minMaxLoc(roi, 0, 0, 0, &maxPoint);
				maxPoint.x = maxPoint.x + i - block_size;
				maxPoint.y = maxPoint.y + j - block_size;
				if (i == maxPoint.x && j == maxPoint.y){
					edge.y_edge.push_back(maxPoint);
					//rectangle(img, maxPoint - Pblock_size, maxPoint + Pblock_size, cvScalar(255, 0, 0), 2);
				}
			}
		}
	}
	//cout << edge.x_edge.size()<<"shuliang"<<edge.y_edge.size();
	return edge;
}